cloud/deploy/deploy.go

```diff
@@ -807,6 +807,101 @@ func WarnIfNonLatestVersion(version string, httpClient *httputil.HTTPClient) {
 	}
 }
 
+// ClientBuildContext represents a prepared build context for client deployment
+type ClientBuildContext struct {
+	// TempDir is the temporary directory containing the build context
+	TempDir string
+	// CleanupFunc should be called to clean up the temporary directory
+	CleanupFunc func()
+}
+
+// prepareClientBuildContext creates a temporary build context with client dependency files
+// This avoids modifying the original project files, preventing race conditions with concurrent deployments.
+func prepareClientBuildContext(sourcePath string) (*ClientBuildContext, error) {
+	// Create a temporary directory for the build context
+	tempBuildDir, err := os.MkdirTemp("", "astro-client-build-*")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create temporary build directory: %w", err)
+	}
+
+	// Cleanup function to be called by the caller
+	cleanup := func() {
+		os.RemoveAll(tempBuildDir)
+	}
+
+	// Check if source directory exists first
+	if exists, err := fileutil.Exists(sourcePath, nil); err != nil {
+		cleanup()
```

Comment:
I think we can make this simpler by having the caller handle this cleanup entirely, instead of having to call it for each error case here, as well as having the caller handle it

Reply:
I thought the same, but no, we can't. Because there's nothing to clean up in the case of the first error here:
```
	// Create a temporary directory for the build context
	tempBuildDir, err := os.MkdirTemp("", "astro-client-build-*")
	if err != nil {
		return nil, fmt.Errorf("failed to create temporary build directory: %w", err)
	}
```
I was also about to blame the AI, but then realised that this is correct.